pyrovision/datasets/utils.py

```diff
@@ -3,153 +3,34 @@
-                        success = True
-                    except Exception:
-                        success = False
-                # Handle exception
-                if not success and (idx == retries):
-                    if not silent:
-                        raise e
-                    elif verbose:
-                        print(e)
-            if success:
-                break
-
-
-def parallel(func, arr, threads=None, leave=False):
-    """Download a file accessible via URL with mutiple retries
-
-    Args:
-        func (callable): function to be executed on multiple workers
-        arr (iterable): function argument's values
-        threads (int, optional): number of workers to be used for multiprocessing
-        leave (bool, optional): whether traces of progressbar should be kept upon termination
-
-    Returns:
-        list: list of function's results
-    """
-
-    if threads is None:
-        threads = min(16, mp.cpu_count())
-    if threads < 2:
-        results = [func(arg) for arg in tqdm(arr, total=len(arr), leave=leave)]
-    else:
-        with ThreadPool(threads) as tp:
-            results = list(tqdm(tp.imap_unordered(func, arr), total=len(arr)))
-    if any([o is not None for o in results]):
-        return results
-
-
-def download_urls(entries, root, timeout=4, retries=4, threads=None, silent=True):
-    """Download multiple URLs a file accessible via URL with mutiple retries
-
-    Args:
-        entries (list<str, str>): URL and destination filen
-        root (pathlib.Path): folder where the files will be saved in
-        timeout (float, optional): number of seconds before the request times out
-        retries (int, optional): number of additional allowed download attempts
-        threads (int, optional): number of threads to be used for multiprocessing
-        silent (bool, optional): whether Exception should be raised upon download failure
-    """
-
-    parallel(partial(download_url, root=root, timeout=timeout, retries=retries, silent=silent),
-             entries, threads=threads)
+import zipfile
+
+
+class DownloadProgressBar(tqdm):
+    """Progress bar."""
+
+    def update_to(self, b=1, bsize=1, tsize=None):
+        if tsize is not None:
+            self.total = tsize
+        self.update(b * bsize - self.n)
+
+
+def download_dataset_from_url(url, root):
+    """Download datatset."""
+    output_path = url.split('/')[-1]
+    with DownloadProgressBar(unit='B', unit_scale=True,
+                             miniters=1, desc=url.split('/')[-1]) as t:
+        urllib.request.urlretrieve(url, filename=output_path, reporthook=t.update_to)
+
+        # check integrity of downloaded file
+        with open(output_path, 'rb') as f:
+            sha_hash = hashlib.sha256(f.read()).hexdigest()
+        if sha_hash[:8] != url.split('-')[-1].split('.')[0]:
+            raise RuntimeError("File not found or corrupted.")
```

Comment:
Downloading & extracting are two independent things: I suggest updating the previous function for download, and if you need to compose them, adding a function calling the latter :ok_hand:

Reply:
Let's leverage existing functions from torchvision if possible : https://github.com/pytorch/vision/blob/master/torchvision/datasets/utils.py